lib/membrane_h264_plugin/parser.ex

```diff
@@ -276,34 +435,109 @@ defmodule Membrane.H264.Parser do
 
-      {sps_actions ++ buffers_actions, state}
+      {stream_format_actions ++ buffers_actions, state}
     end)
   end
 
-  defp maybe_parse_sps(au, state) do
-    case Enum.find(au, &(&1.type == :sps)) do
-      nil ->
-        {[], state}
+  @spec process_new_parameter_sets([NALu.t()], [NALu.t()], CallbackContext.t(), state()) ::
+          {[Action.t()], state()}
+  defp process_new_parameter_sets(new_spss, new_ppss, context, state) do
+    updated_cached_spss = merge_parameter_sets(new_spss, state.cached_spss, :seq_parameter_set_id)
+    updated_cached_ppss = merge_parameter_sets(new_ppss, state.cached_ppss, :pic_parameter_set_id)
+
+    state = %{state | cached_spss: updated_cached_spss, cached_ppss: updated_cached_ppss}
+
+    latest_sps = List.last(new_spss)
+
+    last_sent_stream_format = context.pads.output.stream_format
+
+    output_raw_stream_structure =
+      case state.output_stream_structure do
+        :annexb ->
+          :annexb
+
+        {avc, _nalu_length_size} = output_stream_structure ->
+          {avc,
+           DecoderConfigurationRecord.generate(
+             Enum.map(updated_cached_spss, fn {_id, sps} -> sps.payload end),
+             Enum.map(updated_cached_ppss, fn {_id, pps} -> pps.payload end),
+             output_stream_structure
+           )}
+      end
+
+    stream_format_candidate =
+      case {latest_sps, last_sent_stream_format} do
+        {nil, nil} ->
+          nil
+
+        {nil, last_sent_stream_format} ->
+          %{last_sent_stream_format | stream_structure: output_raw_stream_structure}
+
+        {latest_sps, _last_sent_stream_format} ->
+          Format.from_sps(latest_sps, output_raw_stream_structure,
+            output_alignment: state.output_alignment
+          )
+      end
 
-      sps_nalu ->
-        fmt = Format.from_sps(sps_nalu, output_alignment: state.output_alignment)
-        {[stream_format: {:output, fmt}], %{state | profile: fmt.profile}}
+    if stream_format_candidate in [last_sent_stream_format, nil] do
+      {[], state}
+    else
+      {
+        [stream_format: {:output, stream_format_candidate}],
+        %{state | profile: stream_format_candidate.profile}
+      }
     end
   end
 
+  @spec merge_parameter_sets([NALu.t()], %{non_neg_integer() => NALu.t()}, atom()) ::
+          %{non_neg_integer() => NALu.t()}
+  defp merge_parameter_sets(new_parameter_sets, cached_parameter_sets, id_key) do
+    new_parameter_sets
+    |> Enum.map(&{&1.parsed_fields[id_key], &1})
+    |> Map.new()
+    |> then(&Map.merge(cached_parameter_sets, &1))
+  end
+
+  @spec process_au_parameter_sets(AUSplitter.access_unit(), CallbackContext.t(), state()) ::
+          {AUSplitter.access_unit(), [Action.t()], state()}
+  defp process_au_parameter_sets(au, context, state) do
+    au_spss = Enum.filter(au, &(&1.type == :sps))
+    au_ppss = Enum.filter(au, &(&1.type == :pps))
+
+    {stream_format_actions, state} = process_new_parameter_sets(au_spss, au_ppss, context, state)
```

Comment:
is it guaranteed, that `au_spss` and `au_ppss` passed to the `process_new_parameter_sets` are indeed "new_parameter_sets"?